renku_notebooks/api/classes/crc.py

```diff
@@ -0,0 +1,147 @@
+            )
+        resource_pools = res.json()
+        pool = None
+        res_class = None
+        for rp in resource_pools:
+            for cls in rp["classes"]:
+                if cls["id"] == class_id:
+                    res_class = cls
+                    pool = rp
+        if res_class is None:
+            raise InvalidComputeResourceError(
+                message=f"The resource class ID {class_id} does not exist."
+            )
+        if storage is None:
+            storage = res_class.get("default_storage", 1)
+        if storage < 1:
+            raise InvalidComputeResourceError(
+                message="Storage requests have to be greater than or equal to 1GB."
+            )
+        if storage > res_class.get("max_storage"):
+            raise InvalidComputeResourceError(
+                message="The requested storage surpasses the maximum value allowed."
+            )
+        # Memory and disk space in CRC are assumed to be in gigabytes whereas
+        # the notebook service assumes that if a plain number is used then it is bytes.
+        options = ServerOptions.from_resource_class(res_class)
+        options.set_storage(storage, gigabytes=True)
+        quota = pool.get("quota")
+        if quota is not None and isinstance(quota, dict):
+            options.priority_class = quota.get("id")
+        return options
+
+    def get_default_class(self) -> Dict[str, Any]:
+        res = requests.get(self.crc_url + "/resource_pools")
+        if res.status_code != 200:
+            raise IntermittentError(
+                "The CRC sent an unexpected response, please try again later."
+            )
+        pools = res.json()
+        default_pools = [p for p in pools if p.get("default", False)]
+        if len(default_pools) < 1:
+            raise ConfigurationError("Cannot find the default resource pool.")
+        default_pool = default_pools[0]
+        default_classes = [
+            cls for cls in default_pool.get("classes", []) if cls.get("default", False)
+        ]
+        if len(default_classes) < 1:
+            raise ConfigurationError("Cannot find the default resource class.")
+        return default_classes[0]
+
+    def find_acceptable_class(
+        self, user: User, requested_server_options: ServerOptions
+    ) -> Optional[ServerOptions]:
+        """Find a resource class that is available to the user that is greater than or equal to
+        the old-style server options that the user requested."""
+        headers = None
+        if user.access_token is not None:
+            headers = {"Authorization": f"bearer {user.access_token}"}
+        res = requests.get(self.crc_url + "/resource_pools", headers=headers)
+        if res.status_code != 200:
+            raise IntermittentError(
+                message="The compute resource access control service sent "
+                "an unexpected response, please try again later",
+            )
+        resource_pools = res.json()
```

Comment:
This code is duplicated three times (here and above), maybe it should be its own method?